### PR TITLE
Adding FAISS NN plugin

### PR DESCRIPTION
**Pull request by fishcorn**
Also adding a simple random projections plugin (simple_rp.py).

### CODE REVIEW

**Comment by fishcorn** [comment]
Reopening to restart build

File: python/smqtk/algorithms/nn_index/faiss.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        """
+        Initialize MRPT index properties. Does not contain a queryable index
+        until one is built via the ``build_index`` method, or loaded from
+        existing model files.
+
+        :param descriptor_set: Index in which DescriptorElements will be
+            stored.
+        :type descriptor_set: smqtk.representation.DescriptorIndex
+
+        :param index_filepath: Optional file location to load/store MRPT index
```
Comment (Purg) [inline python/smqtk/algorithms/nn_index/faiss.py]: This parameter doesn't seem to exist.

File: python/smqtk/algorithms/nn_index/faiss.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            when initialized and/or built.
+
+            If not configured, no model files are written to or loaded from
+            disk.
+        :type index_filepath: None | str
+
+        :param parameters_filepath: Optional file location to load/save index
```
Comment (Purg) [inline python/smqtk/algorithms/nn_index/faiss.py]: This parameter doesn't seem to exist.

File: python/smqtk/algorithms/nn_index/faiss.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        self.read_only = read_only
+        self.use_multiprocessing = use_multiprocessing
+        self.pickle_protocol = pickle_protocol
+        self.exhaustive = exhaustive
+        self._descriptor_set = descriptor_set
+
+        def normpath(p):
+            return (p and osp.abspath(osp.expanduser(p))) or p
+
+        # self._index_filepath = normpath(index_filepath)
```
Comment (Purg) [inline python/smqtk/algorithms/nn_index/faiss.py]: Remove commented out code.

File: python/smqtk/algorithms/nn_index/faiss.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            collections.Iterable[smqtk.representation.DescriptorElement]
+
+        """
+        if self.read_only:
+            raise ReadOnlyError("Cannot modify container attributes due to "
+                                "being in read-only mode.")
+
+        super(FaissNearestNeighborsIndex, self).build_index(descriptors)
+
+        self._log.info("Building new MRPT index")
```
Comment (Purg) [inline python/smqtk/algorithms/nn_index/faiss.py]: This is not MRPT

File: python/smqtk/algorithms/nn_index/lsh/functors/simple_rp.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+from smqtk.utils.bin_utils import report_progress
+
+
+class SimpleRPFunctor (LshFunctor):
+    '''
+    This class is meant purely as a baseline comparison for other 
+    LshFunctors and NNIndex plugins. It is not meant to be used in 
+    production, as it is unlikely to produce a quality index.
+    '''
+
```
Comment (Purg) [inline python/smqtk/algorithms/nn_index/lsh/functors/simple_rp.py]: Remove extra new-line.

File: python/smqtk/algorithms/nn_index/faiss.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+        self._log.debug("Clearing and adding new descriptor elements")
+        self._descriptor_set.clear()
+        self._descriptor_set.add_many_descriptors(descriptors)
+
+        self._log.debug('Building FAISS index')
+        self._build_faiss_model()
+
+        # self._save_faiss_model()
+
```
Comment (Purg) [inline python/smqtk/algorithms/nn_index/faiss.py]: Remove extra new-line.

File: python/smqtk/algorithms/nn_index/lsh/functors/simple_rp.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        self.mean_vec = np.mean(x, axis=0, keepdims=True)
+
+        self._log.debug("Generating random projections")
+        np.random.seed(self.random_seed)
+        self.rps = np.random.randn(dim, self.bit_length)
+
+        # self._log.debug("Info normalizing descriptors with norm type: %s",
+        #                 self.normalize)
+        # x = self._norm_vector(x)
+        # return (x >= 0.0)
```
Comment (Purg) [inline python/smqtk/algorithms/nn_index/lsh/functors/simple_rp.py]: Missing return stated in docstring.

File: python/smqtk/algorithms/nn_index/lsh/functors/simple_rp.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        return {
+            "bit_length": self.bit_length,
+            "normalize": self.normalize,
+            "random_seed": self.random_seed,
+        }
+
+    def has_model(self):
+        return (self.mean_vec is not None)
+
+    def fit(self, descriptors, use_multiprocessing=True):  #
```
Comment (Purg) [inline python/smqtk/algorithms/nn_index/lsh/functors/simple_rp.py]: Remove trailing ``#``

**Pull request merged**